cpp/src/fil/infer.cu

```diff
@@ -260,21 +301,36 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_FEW_CLASSES> : finalize_block {
     }
     __syncthreads();  // free up per_thread[] margin
 
-    write_best_class_in_block(to_vec(threadIdx.x, acc), num_classes, out,
-                              num_rows);
+    int c = threadIdx.x;
+    if (num_outputs == 1) {  // will output class
+      write_best_class_in_block(to_vec(c, acc), num_classes, out, num_rows);
+    } else {  // output softmax-ed margin
+      // softmax phase 1: compute sum of exponents
+      auto max = block_all_max(acc, num_classes);
```

Comment:
Would it be possible to separate softmax computation from writing out the result?